### ISSUE TITLE
Thoughts about profiling

### ISSUE THREAD

**Issue by kmcnaught** [OptiKey/OptiKey#819]
This is a place to dump thoughts about possible ways to improve Optikey's performance in the future

- I believe the "look to scroll" and "gaze overlay" features have added some slowdown. 
- A big factor when investigating was many calls to `Graphics.DpiX` etc. This feels like something that could be cached (and potentially subscribed to changes, or re-computed upon certain events)
- When switching between dynamic keyboards that are docked, if they have a different size then there is a fair amount of repositioning lag. This may be unavoidable but it would be worth investigating whether it's possible to resize the app bar without as many calls.

**Comment by AdamRoden** [comment]
I've tried improving the re-positioning lag, it happens even from one floating size to another. The annoying thing is when you see your new layout displayed before the window gets resized. Do you know what I mean?

**Comment by hyperdeath666** [comment]
@AdamRoden Yes, and to add, this can cause input errors for extremely low completion times. Not the end of the world and I have some workarounds, but improvements would be lovely if feasible.

**Comment by kmcnaught** [comment]
@AdamRoden with floating keyboards, is the lag worse between dynamic keyboards compared to built-in ones? I thought that was the case until I started writing this issue, and then couldn't replicate. I definitely recognise your description of the layout appearing early though. 

@hyperdeath666 good point - I have sometimes considered whether the dwell progress should be paused until the layout is complete, but speeding things up would help too.

**Comment by hyperdeath666** [comment]
@kmcnaught That's really interesting. Could make for a great setting?

**Comment by AdamRoden** [comment]
@kmcnaught I get about the same lag regardless of whether going from a built-in to dynamic or vice versa. In the code where Optikey resizes itself, it uses an ApplicationIdle dispatcher which is asynchronous, low priority, and results in the keyboard loading before the window changes size. If you increase the dispatcher priority you can get the window to resize immediately, but then you can end up with your old layout in the new size for a split second. So, we can re-prioritize things for different behavior, but the bottom line is that it takes a split second to load a layout and another split second to resize a window. Now, since the current state does the two asynchronously, maybe we need to re-evaluate that aspect.
